AdblockPlusSafariExtension/AdblockPlus+Parsing.m

```diff
@@ -0,0 +1,308 @@
+}
+
+static int reformat_start_map(void *ctx)
+{
+  adblock_plus_context *context = (adblock_plus_context *)ctx;
+  return yajl_gen_status_ok == yajl_gen_map_open(context->g);
+}
+
+static int reformat_end_map(void *ctx)
+{
+  adblock_plus_context *context = (adblock_plus_context *)ctx;
+  return yajl_gen_status_ok == yajl_gen_map_close(context->g);
+}
+
+static int reformat_start_array(void *ctx)
+{
+  adblock_plus_context *context = (adblock_plus_context *)ctx;
+  context->array_level += 1;
+  return yajl_gen_status_ok == yajl_gen_array_open(context->g);
+}
+
+static int reformat_end_array(void *ctx)
+{
+  adblock_plus_context *context = (adblock_plus_context *)ctx;
+  context->array_level -= 1;
+  if (context->array_level == 0) {
+    return YES;
+  } else {
+    return yajl_gen_status_ok == yajl_gen_array_close(context->g);
+  }
+}
+
+static yajl_callbacks callbacks = {
+  reformat_null,
+  reformat_boolean,
+  NULL,
+  NULL,
+  reformat_number,
+  reformat_string,
+  reformat_start_map,
+  reformat_map_key,
+  reformat_end_map,
+  reformat_start_array,
+  reformat_end_array
+};
+
+static BOOL writeDictionary(NSDictionary<NSString *, id> *__nonnull dictionary, yajl_gen g);
+
+static BOOL writeString(NSString *__nonnull string, yajl_gen g)
+{
+  const char *str = [string cStringUsingEncoding:NSUTF8StringEncoding];
+  return yajl_gen_status_ok == yajl_gen_string(g, str, strlen(str));
+}
+
+static BOOL writeArray(NSArray<id> *array, yajl_gen g)
+{
+  if (yajl_gen_status_ok != yajl_gen_array_open(g)) {
+    return NO;
+  }
+
+  for (id value in array) {
+
+    BOOL result;
+    if ([value isKindOfClass:[NSDictionary class]]) {
+      result = writeDictionary(value, g);
+    } else if ([value isKindOfClass:[NSString class]]) {
+      result = writeString(value, g);
+    } else if ([value isKindOfClass:[NSArray class]]) {
+      result = writeArray(value, g);
+    } else {
+      result = yajl_gen_status_ok == yajl_gen_null(g);
+    }
+
+    if (!result) {
+      return NO;
+    }
+  }
+
+  return yajl_gen_status_ok == yajl_gen_array_close(g);
+}
+
+static BOOL writeDictionary(NSDictionary<NSString *, id> *__nonnull dictionary, yajl_gen g)
+{
+  if (yajl_gen_status_ok != yajl_gen_map_open(g)) {
+    return NO;
+  }
+
+  for (NSString *key in dictionary) {
+    id value = dictionary[key];
+
+    if (!writeString(key, g)) {
+      return NO;
+    }
+
+    BOOL result;
+    if ([value isKindOfClass:[NSDictionary class]]) {
+      result = writeDictionary(value, g);
+    } else if ([value isKindOfClass:[NSString class]]) {
+      result = writeString(value, g);
+    } else if ([value isKindOfClass:[NSArray class]]) {
+      result = writeArray(value, g);
+    } else {
+      result = yajl_gen_status_ok == yajl_gen_null(g);
+    }
+
+    if (!result) {
+      return NO;
+    }
+  }
+
+  return yajl_gen_status_ok == yajl_gen_map_close(g);
+}
+
+
+
+@implementation AdblockPlus (Parsing)
+
++ (NSString *)escapeHostname:(NSString *)hostname
```

Comment:
Yeah, it was used in initial implementation, when whitelisted websites were stored in url-filter. Then QA reported, that whitelisting did not working, so that I moved it to the if-domain. Escaping are no longer needed there.